[Woo POS][Local Catalog] Skip manual one-time sync on site switch if it's not needed

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
WOOMOB-1327

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR refactors the triggerManualFullCatalogSync() method to check sync requirements before triggering a full catalog sync, preventing unnecessary sync operations when the catalog is already up to date or when the local catalog is disabled.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
The feature can be tested by switching back and forth between sites and observing if the full sync is performed.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Skip manual one-time sync if not needed

samiuelson committed:
Merge branch 'trunk' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

malinajirka reviewed (commented):
Thanks for working on this @samiuelson!

I'd like to propose an alternative approach and hear your thoughts about it. I'm thinking it might be easier to follow the code, if we had these checks within the worker instead of having them in the scheduler for two reasons:
1. Single source of truth - skipping sync because the data are up to date is similar to skipping it because POS is inactive, and this logic currently lives in the worker itself.
2. Skipping sync when the check sync requirements fails feels off to me - I believe we should retry it in such case. The worker handles this scenario already, since we run `retry()` on an unexpected error. However, this PR doesn't handle this scenario I believe.


No matter which approach we decide to go with, I have a couple more thoughts:
1. We skip when sync is `NotRequired` => however, `NotRequired` is returned for any site that hasn't been synced in 0-6.999 days. Is this intentional or should we run it even if it has been run more than 24 hours ago?
2. `logger.d("Manual full sync skipped: Catalog is up to date")` feels slightly misleading, I'd suggesting updating the log to `Catalog was updated on xyz.`. Wdyt?

samiuelson committed:
Merge branch 'trunk' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

samiuelson committed:
Move sync requirement check to worker

Moves the logic for checking if a full catalog sync is required from the `WooPosLocalCatalogSyncScheduler` to the `WooPosLocalCatalogSyncWorker`.

This change simplifies the scheduler by making it responsible only for enqueuing the sync work, while the worker itself now determines if the sync should proceed or be skipped based on the current status.

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

samiuelson committed:
Merge branch 'trunk' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

samiuelson committed:
Merge branch 'trunk' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

samiuelson committed:
Pass correct site id

samiuelson committed:
Update tests

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosLocalCatalogSyncWorker.kt:
```diff
@@ -47,6 +48,8 @@ constructor(
 
+        checkSyncRequirement()?.let { return it }
+
         val site = isCatalogSyncSupported() ?: return Result.failure()
```
Is this check still needed when we are using the checker? Seems like there is some duplication in the status checks, wdyt?

samiuelson replied:
Refactored in f0726ab and f95953e

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosLocalCatalogSyncWorker.kt:
```diff
@@ -84,6 +87,33 @@ constructor(
     }
 
+    @Suppress("ReturnCount")
+    private suspend fun checkSyncRequirement(): Result? {
```
It'd be nice to add unit tests for this logic, wdyt?

samiuelson replied:
Added: 12d1651

malinajirka reviewed (approved):
Thanks @samiuelson! It works as expected, I've left a couple non-blocking comments.

samiuelson committed:
Return error instead of throwing in case site is null

samiuelson committed:
Remove the redundant `isLocalCatalogSupported` dependency

samiuelson committed:
Add tests

samiuelson committed:
Merge branch 'trunk' into woomob-1327-woo-poslocal-catalog-do-not-run-full-sync-on-site-switch
